[IMP] sales: dynamic text in pdfs update (#17141)
odoo/documentation: content/applications/sales/sales/sales_quotations/pdf_quote_builder/dynamic_text.rst

.. _Adobe: https://helpx.adobe.com/acrobat/desktop/work-with-pdf-forms/create-forms/convert-to-forms.html

=====================================
Configure dynamic form fields in PDFs
=====================================

When creating custom PDFs to import into the **Sales** app, use *dynamic form fields* to auto-fill
the PDF content with relevant information from the Odoo database, including names and prices.

*Dynamic form fields* are configurable text fields that can be added to a PDF form, and Odoo
automatically fills them with values from the sales order/product associated with the quote.

Formatting tips
===============

When designing the base PDF template for a quote’s header or footer, keep the placements of *dynamic
form field* in mind. Use the following tips to avoid overlapping text and other formatting issues:

* **Leave whitespace**: Ensure there is enough room for dynamic data to expand without overlapping
  logos or borders.
* **Add line breaks**: Place dynamic form fields, like *Customer Name*, on their own lines.
  Alternatively, put them at the end of phrases to avoid text breaking. Long names can push static
  text out of alignment.

> * **Set a default font**: Configure a universally available default font to maintain a consistent

Reviewer: it looks like the * renders fine into a bullet point, but the RST cheatsheet uses dashes as an example

```suggestion
- **Leave whitespace**: Ensure there is enough room for dynamic data to expand without overlapping
  logos or borders.
- **Add line breaks**: Place dynamic form fields, like *Customer Name*, on their own lines.
  Alternatively, put them at the end of phrases to avoid text breaking. Long names can push static
  text out of alignment.
```

information in Odoo.

When configuring the :guilabel:`Path`, use the dot (`.`) notation to specify relationships between
models. Headers and footers start from the current :guilabel:`sale_order` model. Product documents
follow their path from :guilabel:`sale_order_line`.

 .. example::
   A company wants to display the customer's country in its quotations. To display the customer's
   country in the quotation PDF, they used the dynamic form field name
   :guilabel:`invoice_partner_country` in the PDF template. After uploading the PDF to the **Sales**
   app, they set the :guilabel:`Path` of the :guilabel:`Form Field Name` to:

   - For a header or footer document: :guilabel:`partner_invoice_id.country_id.name`

   .. figure:: dynamic_text/example-country-quotation-path.png
      :alt: The customer's country form field pathing for a quotation.

      Example of Odoo's Form Field Path for the customer's country from a quotation.

   - For a product document: :guilabel:`order_id.partner_invoice_id.country_id.name`

   .. figure:: dynamic_text/example-country-product-document-path.png
      :alt: The customer's country form field pathing from the product document.

>       Example of Odoo's Form Field Path for the customer's country from a product document.

Reviewer: the alignment is off by one space. The content within the example block must be aligned with the "e" in `.. example`, and the `..example` is not fully aligned with the beginning of the line.

Separately, the figure and the captions can be aligned with the bullet point

notice the difference? (: 

Current version:

<img width="891" height="529" alt="Image" src="https://github.com/user-attachments/assets/20e6966e-8801-480b-bbdb-7d50b0514a50" />

Revised alignment:

<img width="930" height="697" alt="Image" src="https://github.com/user-attachments/assets/53ae81af-4c45-49e4-9ccc-93c53d471f0b" />

```suggestion
.. example::

    A company wants to display the customer's country in its quotations. To display the customer's
    country in the quotation PDF, they used the dynamic form field name
    :guilabel:`invoice_partner_country` in the PDF template. After uploading the PDF to the
    **Sales** app, they set the :guilabel:`Path` of the :guilabel:`Form Field Name` to:

    - For a header or footer document: :guilabel:`partner_invoice_id.country_id.name`

      .. figure:: dynamic_text/example-country-quotation-path.png
         :alt: The customer's country form field pathing for a quotation.

         Example of Odoo's Form Field Path for the customer's country from a quotation.

    - For a product document: :guilabel:`order_id.partner_invoice_id.country_id.name`

      .. figure:: dynamic_text/example-country-product-document-path.png
         :alt: The customer's country form field pathing from the product document.

         Example of Odoo's Form Field Path for the customer's country from a product document.
```

.. _Adobe: https://helpx.adobe.com/acrobat/desktop/work-with-pdf-forms/create-forms/convert-to-forms.html

================================
Configuring dynamic text in PDFs

> ================================

Reviewer 2: ```suggestion
Configure dynamic form fields in PDFs
================================
```

Reviewer 2: present tense to match other articles; simpler than gerund

While creating custom PDFs for quotes, use *dynamic text* for Odoo to auto-fill the PDF content with
information related to the quote from the Odoo database, like names, prices, etc.

Dynamic text values are form components (text inputs) that can be added in a PDF file, and Odoo

> automatically fills those values in with information related to the quote.

Reviewer 2: ```suggestion
*Dynamic form fields* are components that can be added to a PDF form, and Odoo
automatically fills those text inputs with values from the sales order/product related to the quote.
```

Reviewer 2: one term, whichever you pick

specify _pdf form_ -- using that term bc it's used in the adobe article

Design tips
===========

> When designing the base PDF template for a quote’s header or footer, keep the dynamic text placement

Reviewer 2: > When designing the base PDF template for a quote’s header or footer

does this not apply to `product document` type pdfs?

in mind. Use the following tips to avoid overlapping text and poor design:

* **Leave Whitespace**: Ensure there is enough room for dynamic data to expand without overlapping
  logos or borders.
* Place dynamic fields, like Customer Name, on their own lines. Alternatively, put them at the end
  of phrases to avoid text breaking. Long names can push static text out of alignment.
* **Set a default font**: Configuring a universally available default font maintains a consistent
  text appearance when adding or editing content.

Edit PDF forms with Adobe software. Form fields in header and footer templates are required to

> retrieve dynamic values in Odoo.

Reviewer 2: also consider flipping sentence to start w/intent "to retrieve dynamic values"

Scribus.

Adobe Acrobat Pro
-----------------

First, `convert the PDF into a PDF form <Adobe_>`_. Then add a text field at the location where the
Odoo information needs to be displayed. Link the Odoo variables to the text fields by
double-clicking on the field to open :guilabel:`Properties`.

In the :guilabel:`General tab`, enter the Odoo field variable for the :guilabel:`Name`. Set the
:guilabel:`Common Properties` to :guilabel:`Visible` unless the field needs to be hidden until data
is populated.

Next, click the :guilabel:`Appearance` tab and select the :guilabel:`Font Size`, :guilabel:`Font
Choice`, and :guilabel:`Text Color` to match the template's existing text or branding. Click the
:guilabel:`Options` tab and set the text alignment to match the template’s design.

.. tip::
   Refer to the :ref:`pdf_quote_builder/dynamic_text/common-dynamic-text-values` section for usual
   Odoo field variables and how to find a specific field value in Odoo.

General PDF editor instructions
-------------------------------

Open the desired PDF in the chosen PDF editor app. Then add a text field at the location where the

> Odoo information needs to be displayed. Link the Odoo variables to the text fields by opening the

Reviewer 2: ```suggestion
Odoo information should be inserted. Link the Odoo fields to the *dynamic form fields* by opening the
```

field’s :guilabel:`Properties` window. Then, in the :guilabel:`Name` for that field, enter the Odoo

> variable.

Reviewer 2: ```suggestion
field’s :guilabel:`Properties` window. Then, in the :guilabel:`Name` for that field, enter the name of
the Odoo field.
```

If possible, configure the :guilabel:`Font Size`, :guilabel:`Font Choice`, and :guilabel:`Text
Color` to match the template's existing text or branding. Click the :guilabel:`Options` tab and set
the text alignment to match the template’s design.

> Link PDF text fields to Odoo fields

Reviewer 2: to make it clear this is one of 2 configuration sections, and not a discrete/separate topic vs. _prepare the dynamic text fields_

===================================

Once the PDF file(s) are complete, save them to the computer's hard drive, and proceed to upload
them to Odoo via :menuselection:`Sales app --> Configuration --> Headers/Footers`.

> If the uploaded PDF contains configured text forms, Odoo will automatically detect them and display

Reviewer 2: ```suggestion
Odoo will automatically detect *dynamic form fields* in the uploaded PDF and display
```

> the :guilabel:`Configure dynamic fields` link. This step connects any field name in the PDF to the

Reviewer 2: > This step connects

feel like this should be the first sentence in the section since it explains the point of the section

> correct Odoo field by specifying the exact technical path (the database code) to the desired

Reviewer 2: ```suggestion
corresponding Odoo field by specifying the exact location of that
```

information.

Click the :guilabel:`Configure dynamic fields` link to go to the :guilabel:`Form Fields` page. The
:guilabel:`Form Field Name` column is populated with the text fields from the PDF. The
:guilabel:`Path` column is where the Odoo field path needs to be specified for each form field.

.. image:: dynamic_text/form-fields-page.png
   :alt: Configured text form fields in the PDF quote builder.

> To edit, click in the :guilabel:`Path` cell of the desired form field row and enter the dynamic text

Reviewer 2: consider making a _how to find/add path_ subsection for _odoo config_

value. Refer to the :ref:`pdf_quote_builder/dynamic_text/common-dynamic-text-values` section for
usual Odoo field variables and how to find a specific field value in Odoo.

.. note::

>    Headers and footers starts from the current :guilabel:`sale_order` model. Product documents

Reviewer 2: consider line break so the two sentences are parallel above/below each other

   dynamic text field, the Sales Order Reference in the :guilabel:`name` field, and the Customer
   Name in the :guilabel:`partner_id_name` field.

   .. image:: dynamic_text/pdf-quote-builder-sample.png
      :align: center
      :alt: PDF quote being built using common dynamic placeholders.

.. example::
   When uploading PDF containing the form field :guilabel:`invoice_partner_country`, which is an
   information available in the sales order, configure the :guilabel:`path` of the :guilabel:`Form
   Field Name` to:

   - :guilabel:`partner_invoice_id.country_id.name` for a header or footer document

   .. figure:: dynamic_text/example-country-quotation-path.png
      :alt: The customer's country form field pathing for a quotation.

      Example of Odoo's Form Field Path for the customer's country from a quotation.

   - :guilabel:`order_id.partner_invoice_id.country_id.name` for a product document fills the form
     with the invoice partner country's name when the PDF is built.

   .. figure:: dynamic_text/example-country-product-document-path.png
      :alt: The customer's country form field pathing from the product document.

>       Example of Odoo's Form Field Path for the customer's country from a product document.

Reviewer 2: this example should go with the `note` admonition about path for _header/footer_ vs. _product documents_ on line 92
